src/vmm/src/devices/virtio/iovec.rs

```diff
@@ -37,19 +37,23 @@ type IoVecVec = SmallVec<[iovec; 4]>;
 /// It describes a buffer passed to us by the guest that is scattered across multiple
 /// memory regions. Additionally, this wrapper provides methods that allow reading arbitrary ranges
 /// of data from that buffer.
-#[derive(Debug)]
+#[derive(Debug, Default)]
 pub struct IoVecBuffer {
     // container of the memory regions included in this IO vector
     vecs: IoVecVec,
     // Total length of the IoVecBuffer
     len: u32,
 }
 
+// SAFETY: `IoVecBuffer` doesn't allow for interior mutability and no shared ownership is possible
+// as it doesn't implement clone
+unsafe impl Send for IoVecBuffer {}
+
 impl IoVecBuffer {
     /// Create an `IoVecBuffer` from a `DescriptorChain`
-    pub fn from_descriptor_chain(head: DescriptorChain) -> Result<Self, IoVecError> {
-        let mut vecs = IoVecVec::new();
-        let mut len = 0u32;
+    pub unsafe fn load_descriptor_chain(&mut self, head: DescriptorChain) -> Result<(), IoVecError> {
+        self.vecs.clear();
```

Comment:
I think we really need an additional `pub fn clear(&mut self) { self.vecs.clear() }` function in these iovecbuffer types that call inside `process_tx` at the very end of the function. That's because for safety, we rely on no two buffers pointing towards the same slices of guest memory. We know this will be true because all buffers across all devices are processed sequentially, but we need to make sure that some cached buffer (such as the one in the net device) does not contain "stale" points while we're processing another buffer. Tbh, this is mainly just me being paranoid, I don't think anything bad can happen from it (because the only thing we do with those stale pointers is overwrite them later), but let's be precise here :)

Reply:
Yeah, sorry, I meant that in addition to clearing the buffer here, it should also get cleared at the end `process_tx`. Because right now what can happen is 
1. we process a TX packet and load the descriptor chain into the cached buffer
2. guest reuses memory location for a vsock package
3. vsock code loads the same memory location into its iovecbuffer, meaning we now have two buffers pointing to the same memory
4. cached buffer in net device is only cleared when the next TX packet arrives.

Reply:
oh, you are right! I had not thought of this scenario. That makes perfect sense. I wonder if there is a way to structure the code so that this is guaranteed at compile time.